### PR TITLE
[RDY] Fix buffer overflow of played_sound_callbacks_ids

### PR DESCRIPTION
**Pull request by mugmuggy**
*Fixes #* 

**Describe what the proposed change does**
- Fix buffer overflow 
- Potentially addresses #1380, was also some older bugs (still bugs?) where lava hole spawns but not accessible by Grim, which would also accelerate the overflow as the lava hole stays spawned.

### CODE REVIEW

**Comment by lewri** [comment]
Have you been able to test any of the possible affected bugs? I don't think we've had any sight of them in years so likely they were fixed or very rare.

**Comment by mugmuggy** [comment]
Grim issue was resolved in #1708. There was another also resolved which was resolved I believe, kicking dying patients as they were on the way to lava hole. I've got the comment on #1380 about a crash with audio. I do recall the function which failed and seen that same buffer overflow cause the same access violation that I seen, but if it addresses 1380 or not I don't know. I think we've since silenced (?) announcements whilst paused, which is where I believe my crash was in 1380, and it was a silent crash to desktop on latest builds.

Its easy enough to replicate, spawn a lava hole through debug console, and you can monitor Audio:playSound for the value of
```lua
self.unused_played_callback_id = self.unused_played_callback_id + 1
```
This value will exceed 1000, get to somewhere else and then overwrites sound_player eventually. The do it twice thing from discord also makes getting to 1000+ quicker

**Comment by lewri** [comment]
> I think we've since silenced (?) announcements whilst paused,

Announcements are halted now unless they had priority critical (so cheats usage was made known)

File: CorsixTH/Src/th_lua_sound.cpp [hunk 1]
```diff
@@ -31,7 +31,8 @@ SOFTWARE.
 
 namespace {
 
-int played_sound_callback_ids[1000];
+constexpr int max_callback_ids = 1000;
```
Comment (TheCycoONE) [inline CorsixTH/Src/th_lua_sound.cpp:34]: Prefer modernizing to std::array<int, 1000> played_sound_callback_ids;

File: CorsixTH/Src/th_lua_sound.cpp [hunk 2]
```diff
@@ -230,7 +231,7 @@ int l_soundfx_play(lua_State* L) {
     if (!lua_isnil(L, 7))
       iPlayedCallbackDelay = static_cast<int>(luaL_checknumber(L, 7));
 
-    if (played_sound_callback_index == sizeof(played_sound_callback_ids))
+    if (played_sound_callback_index == max_callback_ids)
```
Comment (TheCycoONE) [inline CorsixTH/Src/th_lua_sound.cpp:234]: After making this a std::array, len can be used here.

**Review by TheCycoONE — CHANGES_REQUESTED** [review]
We're moving to c++11 and standard container types where possible to prevent overflows. This is a good opportunity to do that instead.

**Review by TheCycoONE — APPROVED** [review]

**Pull request merged**